craft_parts/parts.py

```diff
@@ -0,0 +1,243 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Definitions and helpers to handle parts."""
+
+import copy
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Set
+
+from craft_parts import errors
+from craft_parts.dirs import ProjectDirs
+from craft_parts.steps import Step
+
+
+class Part:
+    """Each of the components used in the project specification.
+
+    During the craft-parts lifecycle each part is processed through
+    different steps in order to obtain its final artifacts. The Part
+    class holds the part specification data and additional configuration
+    information used during step processing.
+
+    :param name: The part name.
+    :param data: A dictionary containing the part properties.
+    :param project_dirs: The project work directories.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        data: Dict[str, Any],
```

Comment:
I know it's probably asking too much right now, but I would recommend avoiding the data/dict pattern for the init.  Enumerate the properties with their sane defaults in the constructor.  add a @classmethod to unmarshal a dict (and perform any required schema/type checking).

Reply:
The reason the original set of properties is kept in dictionary form is because state management filters properties of interest from a dict of properties, for each step (like snapcraft does). Except for that use, the class protects internal data using read-only accessors so refactoring it to get rid of the dict pattern and have all data stored in their own attributes shouldn't be a problem -- but we need a better solution for state building first.

Reply:
I like where your head is at.